test/gtest/tl/mlx5/test_tl_mlx5_wqe.cc

```diff
@@ -0,0 +1,312 @@
+    uint32_t bytes_count           = std::get<1>(GetParam());
+    uint32_t repeat_count          = std::get<2>(GetParam());
+    uint32_t bytes_skip            = std::get<3>(GetParam());
+    int      src_size              = (bytes_count + bytes_skip) * repeat_count;
+    int      dst_size              = bytes_count * nbr_srcs * repeat_count;
+    int      send_mem_access_flags = 0;
+    int      recv_mem_access_flags =
+        IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE;
+    DT                           src[nbr_srcs][src_size], dst[dst_size];
+    struct ibv_mr *              src_mr[nbr_srcs], *dst_mr, *umr_entries_mr;
+    struct mlx5dv_mkey *         umr_mkey;
+    struct mlx5dv_mkey_init_attr umr_mkey_init_attr;
+    void *                       umr_entries_buf;
+    size_t                       umr_buf_size;
+    struct mlx5dv_mr_interleaved mkey_entries[nbr_srcs];
+    struct ibv_wc                wcs[1];
+    int i, j, completions_num, src_index, repetition_count, offset;
+
+    // Setup src and dst buffers
+    for (i = 0; i < nbr_srcs; i++) {
+        for (j = 0; j < src_size; j++) {
+            src[i][j] = (i * (src_size + 7) + j) % 255;
+        }
+        src_mr[i] = ibv_reg_mr(pd, src[i], src_size, send_mem_access_flags);
+        GTEST_ASSERT_NE(nullptr, src_mr[i]);
+    }
+
+    memset(dst, 0, dst_size);
+    dst_mr = ibv_reg_mr(pd, dst, dst_size, recv_mem_access_flags);
+    GTEST_ASSERT_NE(nullptr, dst_mr);
+
+    // UMR
+    umr_buf_size = ucc_align_up(
+        sizeof(struct mlx5_wqe_umr_repeat_ent_seg) * (nbr_srcs + 1), 64);
+    GTEST_ASSERT_EQ(ucc_posix_memalign(&umr_entries_buf, 2048, umr_buf_size),
+                    0);
+
+    umr_entries_mr =
+        ibv_reg_mr(pd, umr_entries_buf, umr_buf_size, send_mem_access_flags);
+    GTEST_ASSERT_NE(nullptr, umr_entries_mr);
+
+    memset(&umr_mkey_init_attr, 0, sizeof(umr_mkey_init_attr));
+    umr_mkey_init_attr.pd           = pd;
+    umr_mkey_init_attr.create_flags = MLX5DV_MKEY_INIT_ATTR_FLAGS_INDIRECT;
+    umr_mkey_init_attr.max_entries  = nbr_srcs + 1; //+1 for the "repeat block"
+
+    umr_mkey = mlx5dv_create_mkey(&umr_mkey_init_attr);
+    GTEST_ASSERT_NE(nullptr, umr_mkey);
+    GTEST_ASSERT_GE(umr_mkey_init_attr.max_entries, nbr_srcs + 1);
+
+    for (i = 0; i < nbr_srcs; i++) {
+        mkey_entries[i].addr        = (uintptr_t)src[i];
+        mkey_entries[i].bytes_count = bytes_count;
+        mkey_entries[i].bytes_skip  = bytes_skip;
+        mkey_entries[i].lkey        = src_mr[i]->lkey;
+    }
+
+    post_umr(umr_qp.qp, umr_mkey, send_mem_access_flags, repeat_count, nbr_srcs,
+             mkey_entries, (uint32_t)umr_entries_mr->lkey, umr_entries_buf);
+
+    completions_num = 0;
+    while (!completions_num) {
+        completions_num = ibv_poll_cq(cq, 1, wcs);
+    }
+    GTEST_ASSERT_EQ(completions_num, 1);
+    GTEST_ASSERT_EQ(wcs[0].status, IBV_WC_SUCCESS);
+    GTEST_ASSERT_EQ(wcs[0].wr_id, 0);
+
+    // RDMA Write
+    ibv_wr_start(qp.qp_ex);
```

Comment:
Why not ibv_wr_start and ibv_wr_complete() part of ucc_tl_mlx5_post_rdma()

Reply:
> Why not ibv_wr_start and ibv_wr_complete() part of ucc_tl_mlx5_post_rdma()

i think because we can post multiple outstanding rmda to the same peer in a batch and make a single critical section for them (it will be in the send_block fn of the algorithm)